### DISCUSSION TITLE
DI: Diferença entre Scoped e Sigleton

### DISCUSSION THREAD

**Discussion by OliveiraCleidson** [balta-io/7196#9] [Perguntas e Respostas] [answered]
Estava vendo a aula do curso 7196 sobre resolução de dependências.

Quando utilizo Scoped ele cria uma instância da classe no momento da requisição, supondo que chegue mais requisições o container entregará está instância já criada e quando todas requisições que estão utilizando está instância terminarem, o container mata a instância e repete o processo quando chegar uma nova requisição, certo? Scoped é utilizado no ciclo de vida das requisições que a utiliza.

Enquanto Sigleton ele cria uma instância e usa ela para resolver as dependências durante todo o ciclo de vida do aplicativo?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia @OliveiraCleidson vou tentar explicar melhor aqui:

**AddSingleton**
Cria uma instância do objeto quando necessário e armazena esta instância em um local na memória. A partir deste momento, toda solicitação para esta instância recebe este mesmo objeto, não é criado um objeto novo.

Isto acontece até sua aplicação ser desligada ou reiniciada. Aí o ciclo se repete.

**AddScopped**
Cria uma instância do objeto e mantém ele ativo durante a requisição, (Thread principal dela). Desta forma, dentro desta requisição, toda chamada a instância, receberá o mesmo objeto, único e exclusivo para esta requisição.

Se sua aplicação receber 200 requisições, ela vai criar 200 objetos, um para cada requisição e manter eles vivos até a requisição terminar.

**AddTransient**
Cria uma instância do objeto para cada chamada ao mesmo, depois de usado (Dispose) o objeto é removido da memória e criado novamente quando preciso.

**Sobre containers**
Os containers rodam de forma independente e não compartilham memória. Portanto, um objeto criado na instância X não é compartilhado com a instância Y.
